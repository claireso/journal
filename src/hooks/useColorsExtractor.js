import { useState, useEffect, useCallback } from 'react'
import ColorThief from 'colorthief'

const rgbToHex = (r, g, b) =>
  '#' +
  [r, g, b]
    .map((x) => {
      const hex = x.toString(16)
      return hex.length === 1 ? '0' + hex : hex
    })
    .join('')

const useColorsExtractor = (initialPhoto) => {
  const [colors, setColors] = useState()

  const extractColors = useCallback(
    (path) => {
      const img = new Image()

      img.src = path

      img.onload = function () {
        const colorThief = new ColorThief()
        const colors = colorThief.getPalette(img).map(([r, g, b]) => rgbToHex(r, g, b))
        setColors(colors)
      }
    },
    [setColors]
  )

  useEffect(() => {
    if (initialPhoto) {
      extractColors(initialPhoto)
    }
  }, [initialPhoto, extractColors])

  return [colors, extractColors]
}

export default useColorsExtractor
