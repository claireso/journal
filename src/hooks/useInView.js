import { useEffect, useState } from 'react'

const config = {
  root: null,
  rootMargin: '0px',
  threshold: 0
}

const useInView = (element) => {
  const [isInView, setInView] = useState(false)

  useEffect(() => {
    const observer = new IntersectionObserver((entries) => {
      entries.forEach((entry) => {
        const _isInView = entry.isIntersecting

        setInView(_isInView)

        if (_isInView) {
          observer.unobserve(entry.target)
        }
      })
    }, config)

    observer.observe(element.current)

    return () => observer.disconnect()
  }, [element])

  return isInView
}

export default useInView
