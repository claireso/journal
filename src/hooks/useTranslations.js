import React from 'react'
import PropTypes from 'prop-types'

import translations from '../translations.json'

const TranslationsContext = React.createContext({})

export const TranslationsProvider = ({ children, translations }) => (
  <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
)

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = () => React.useContext(TranslationsContext)

export const getTranslations = (lang, namespace) => {
  const dict = translations[lang]

  if (!dict) {
    throw new Error(`Language '${lang}' is not supported by the application`)
  }

  if (!namespace) return dict

  const trans = dict[namespace]

  if (!trans) {
    throw new Error(`Namespace '${namespace}' is not defined for the language ${lang}`)
  }

  return trans
}
