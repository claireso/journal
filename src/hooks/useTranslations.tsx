'use client'

import React from 'react'
import PropTypes from 'prop-types'

import { getTranslations } from '../translations'

const TranslationsContext = React.createContext({})

type Translations = {
  [key: string]: object | string
}

interface TranslationsProviderProps {
  children: React.ReactNode
  namespace: 'client' | 'admin'
}

export const TranslationsProvider = ({ children, namespace }: TranslationsProviderProps) => {
  const translations = getTranslations(process.env.NEXT_PUBLIC_WEBSITE_LANGUAGE, namespace)

  return <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
}

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = (): Translations => React.useContext(TranslationsContext)
