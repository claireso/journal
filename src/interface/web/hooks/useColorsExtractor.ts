import { useState, useEffect, useCallback } from 'react'
import ColorThief from 'colorthief'

const rgbToHex = (r: number, g: number, b: number) =>
  '#' +
  [r, g, b]
    .map((x) => {
      const hex = x.toString(16)
      return hex.length === 1 ? '0' + hex : hex
    })
    .join('')

const useColorsExtractor = (initialPhoto?: string) => {
  const [colors, setColors] = useState<string[]>([])

  const extractColors = useCallback(
    (path: string) => {
      const img = new Image()

      img.src = path

      img.onload = function () {
        const colorThief = new ColorThief()
        const colors = colorThief.getPalette(img).map(([r, g, b]) => rgbToHex(r, g, b))
        setColors(colors)
      }
    },
    [setColors]
  )

  useEffect(() => {
    if (initialPhoto) {
      extractColors(initialPhoto)
    }
  }, [initialPhoto, extractColors])

  return [colors, extractColors] as const
}

export default useColorsExtractor
