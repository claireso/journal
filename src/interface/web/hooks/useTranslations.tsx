'use client'

import React from 'react'
import PropTypes from 'prop-types'

import { getTranslations } from '@infrastructure/translations'

const TranslationsContext = React.createContext({})

type Translations = {
  [key: string]: object | string
}

interface TranslationsProviderProps {
  children: React.ReactNode
  namespace: 'client' | 'admin'
  lang?: string
}

export const TranslationsProvider = ({ children, namespace, lang }: TranslationsProviderProps) => {
  const translations = getTranslations(lang ?? 'en', namespace)

  return <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
}

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = (): Translations => React.useContext(TranslationsContext)
