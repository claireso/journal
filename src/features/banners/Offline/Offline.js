import React from 'react'

import { useTranslations } from '@hooks/useTranslations'
import useIsOnline from '@hooks/useIsOnline'

import Flash from '@components/Flash'

const BannerOffline = () => {
  const isOnline = useIsOnline()
  const translations = useTranslations()

  if (isOnline) return null

  return (
    <Flash status="info" css={{ mb: 0 }}>
      {translations.bannerOffline}
    </Flash>
  )
}

export default BannerOffline
