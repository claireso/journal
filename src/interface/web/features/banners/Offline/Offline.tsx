'use client'

import React from 'react'

import { useTranslations } from '@web/hooks/useTranslations'
import useIsOnline from '@web/hooks/useIsOnline'

import Flash from '@web/components/Flash'

const BannerOffline = () => {
  const isOnline = useIsOnline()
  const translations = useTranslations()

  if (isOnline) return null

  return <Flash status="info">{translations.bannerOffline as string}</Flash>
}

export default BannerOffline
