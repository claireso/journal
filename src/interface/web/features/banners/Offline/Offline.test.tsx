// eslint-disable-next-line
// @ts-nocheck
import { render } from '@testing-library/react'

import { TranslationsProvider } from '@web/hooks/useTranslations'
import Offline from './index'

describe('<Offline />', () => {
  const renderBanner = () =>
    render(
      <TranslationsProvider namespace="client">
        <Offline />
      </TranslationsProvider>
    )

  it('should not render component', () => {
    const { asFragment } = renderBanner()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component', () => {
    global.goOffline()

    const { asFragment } = renderBanner()

    expect(asFragment()).toMatchSnapshot()

    global.goOnline()
  })
})
