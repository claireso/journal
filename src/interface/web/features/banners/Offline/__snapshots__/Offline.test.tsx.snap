// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Offline /> should not render component 1`] = `<DocumentFragment />`;

exports[`<Offline /> should render component 1`] = `
<DocumentFragment>
  <div
    class="styles__6gti4e6 styles_layerStyle__6gti4e0 styles_wrapper_status_info__6gti4ea styles_layerStyle__6gti4e4"
  >
    You are currently offline. Connect to the internet for access to the latest photos
  </div>
</DocumentFragment>
`;
